File: python/day2_part1_2023.py

import re

# Read input from file
with open('input.txt', 'r') as file:
    data = file.readlines()

# Define target cube counts
target_cubes = {'red': 12, 'green': 13, 'blue': 14}

# Initialize sum of IDs of possible games
sum_of_ids = 0

# Iterate over each game
for game in data:
    game_id = int(re.search(r'\d+', game).group())
    cubes = re.findall(r'\d+ [a-z]+', game)
    
    game_possible = True
    for cube in cubes:
        count, color = cube.split()
        if target_cubes[color] < int(count):
            game_possible = False
            break
    
    if game_possible:
        sum_of_ids += game_id

print(sum_of_ids)

File: julia/day2_part1_2023.jl

# Function to check if a game is possible with the given cube limits
function is_possible(game_data, limits)
    for subset in game_data
        for (color, count) in subset
            if count > limits[color]
                return false
            end
        end
    end
    return true
end

# Function to parse the input data from the file
function parse_input(filename)
    games = []
    open(filename, "r") do file
        for line in eachline(file)
            parts = split(line, ": ")
            game_id = parse(Int, parts[1][5:end])  # Extract the game ID
            subsets = split(parts[2], "; ")
            game_data = []
            for subset in subsets
                color_counts = Dict{Symbol, Int}()
                for color_info in split(subset, ", ")
                    count, color = split(color_info, " ")
                    color_counts[Symbol(color)] = parse(Int, count)
                end
                push!(game_data, color_counts)
            end
            push!(games, (game_id, game_data))
        end
    end
    return games
end

# Main function to calculate the sum of possible game IDs
function main()
    # Define the limits for the cubes
    limits = Dict(:red => 12, :green => 13, :blue => 14)

    # Parse the input
    games = parse_input("input.txt")

    # Calculate the sum of IDs of possible games
    possible_ids_sum = 0
    for (game_id, game_data) in games
        if is_possible(game_data, limits)
            possible_ids_sum += game_id
        end
    end

    # Print the result
    println("The sum of the IDs of the possible games is: $possible_ids_sum")
end

# Run the main function
main()
